File: cpp/beer-song/beer_song.cpp
#include "beer_song.h"
#include <sstream>

std::string beer::verse(int n) {
  std::stringstream s;
  // handle oddball 0 case separately
  if (n == 0) {
    s << "No more bottles of beer on the wall, no more bottles of beer.\n";
    s << "Go to the store and buy some more, 99 bottles of beer on the wall.\n";
  } else {
    // first line
    std::string bottles = (n == 1 ? "1 bottle" : std::to_string(n) + " bottles");
    s << bottles << " of beer on the wall, " << bottles << " of beer.\n";
    // second line
    std::string object = (n == 1 ? "it" : "one");
    std::string remaining_bottles = (n == 1 ? "no more" : std::to_string(n - 1));
    std::string plurality = (n == 2 ? " bottle " : " bottles ");
    s << "Take " << object << " down and pass it around, " << remaining_bottles << plurality << "of beer on the wall.\n";
  }
  return s.str();
}

std::string beer::sing(int start, int end /*= 0 */) {
  std::string res = "";
  for (int i = start; i >= end; i--) {
    res += verse(i) + "\n";
  }
  return res.substr(0, res.length() - 1);
}

File: java/beer-song/src/main/java/BeerSong.java
class BeerSong {
  String sing(int start, int num) {
    StringBuilder s = new StringBuilder();
    for (int i = start; i > start - num; i--) {
      s.append((i == 0 ? "No more" : i) + (i == 1 ? " bottle" : " bottles") + " of beer on the wall, "
        + (i == 0 ? "no more" : i) + (i == 1 ? " bottle" : " bottles") + " of beer.\n");
      if (i == 0) {
        s.append("Go to the store and buy some more, 99 bottles of beer on the wall.\n\n");
      } else {
        s.append("Take " + (i == 1 ? "it" : "one") + " down and pass it around, "
          + (i == 1 ? "no more" : i - 1) + (i == 2 ? " bottle" : " bottles")
          + " of beer on the wall.\n\n");
      }
    }
    return s.toString();
  }

  String singSong() {
    return sing(99, 100);
  }
}